File: Output/output_TY/c/Medium/845-find-largest-value-in-each-tree-row.c
/**https://leetcode.com/problems/find-largest-value-in-each-tree-row/ */
//Given the root of a binary tree, return an array of the largest value in each row of the tree (0-indexed).
// 
//Example 1:
//Input: root = [1,3,2,5,3,null,9]
//Output: [1,3,9]
//Example 2:
//Input: root = [1,2,3]
//Output: [1,3]
// 
//Constraints:
//	The number of nodes in the tree will be in the range [0, 104].
//	-231 <= Node.val <= 231 - 1
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* largestValues(struct TreeNode* root, int* returnSize){
    int *ret = NULL;
    *returnSize = 0;
    if (NULL == root) {
        return ret;
    }
    int max = root->val;
    struct TreeNode *p = root;
    while (p) {
        if (p->val > max) {
            max = p->val;
        }
        p = p->right;
    }
    ret = (int *)malloc(sizeof(int));
    *ret = max;
    *returnSize = 1;
    return ret;
}

File: Output/output_Codeium/python/Medium/845-find-largest-value-in-each-tree-row.py
#https://leetcode.com/problems/find-largest-value-in-each-tree-row/
#Given the root of a binary tree, return an array of the largest value in each row of the tree (0-indexed).
# 
#Example 1:
#Input: root = [1,3,2,5,3,null,9]
#Output: [1,3,9]
#Example 2:
#Input: root = [1,2,3]
#Output: [1,3]
# 
#Constraints:
#	The number of nodes in the tree will be in the range [0, 104].
#	-231 <= Node.val <= 231 - 1
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def largestValues(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[int]
        """
        if not root:
            return []
        
        from collections import deque
        
        queue = deque([root])
        largest_values = []
        
        while queue:
            level_size = len(queue)
            max_value = float('-inf')
            
            for _ in range(level_size):
                node = queue.popleft()
                max_value = max(max_value, node.val)
                
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            
            largest_values.append(max_value)
        
        return largest_values

